### PR TITLE
Day 21

### PR DESCRIPTION
**Pull request by dmmiller**

### CODE REVIEW

File: 2020/day21/solution.py [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+allergen_map = {}
+
+with open('input.txt') as f:
+    for line in f.readlines():
+        ingredients, allergens = get_ingredients_and_allergens(line.strip())
+        all_ingredients_counter.update(ingredients)
+        for allergen in allergens:
+            if allergen not in allergen_map:
+                allergen_map[allergen] = set(ingredients)
+            allergen_map[allergen] = allergen_map[allergen].intersection(ingredients)
```
Comment (aneeshusa) [inline 2020/day21/solution.py]:
```suggestion
            allergen_map[allergen] &= ingredients
```

File: 2020/day21/solution.py [hunk 2]
```diff
@@ -0,0 +1,43 @@
+from __future__ import annotations
+from collections import Counter
+import re
+
+food_re = re.compile(r"(?P<ingredients>.*) \(contains (?P<allergens>.*)\)")
```
Comment (aneeshusa) [inline 2020/day21/solution.py]:
```suggestion
FOOD_RE = re.compile(r"(?P<ingredients>.*) \(contains (?P<allergens>.*)\)")
```

File: 2020/day21/solution.py [hunk 3]
```diff
@@ -0,0 +1,43 @@
...
+
+def get_ingredients_and_allergens(line: str) -> tuple[set[str], set[str]]:
+    match = food_re.match(line)
+    ingredients = set(ingredient.strip() for ingredient in match['ingredients'].split(' '))
+    allergens = set(ingredient.strip() for ingredient in match['allergens'].split(','))
```
Comment (aneeshusa) [inline 2020/day21/solution.py]:
```suggestion
    ingredients = set(ingredient for ingredient in match['ingredients'].split(' '))
    allergens = set(allergen for allergen in match['allergens'].split(', '))
```
Didn't actually check but seems like you can get rid of the `strip()s` with `','` -> `', '`?

File: 2020/day21/solution.py [hunk 4]
```diff
@@ -0,0 +1,43 @@
...
+                locked[k] = removed_item
+                del allergen_map[k]
+                break
+        for k, v in allergen_map.items():
+            if removed_item in v:
+                v.remove(removed_item)
+
+    return ','.join(locked[key] for key in sorted(locked.keys()))
+    
+all_ingredients_counter = Counter()
```
Comment (aneeshusa) [inline 2020/day21/solution.py]:
```suggestion
ingredients_count = Counter()
```

File: 2020/day21/solution.py [hunk 5]
```diff
@@ -0,0 +1,43 @@
...
+
+with open('input.txt') as f:
+    for line in f.readlines():
+        ingredients, allergens = get_ingredients_and_allergens(line.strip())
+        all_ingredients_counter.update(ingredients)
+        for allergen in allergens:
+            if allergen not in allergen_map:
+                allergen_map[allergen] = set(ingredients)
+            allergen_map[allergen] = allergen_map[allergen].intersection(ingredients)
+    possible_allergens = set(allergen for allergens in allergen_map.values() for allergen in allergens)
```
Comment (aneeshusa) [inline 2020/day21/solution.py]:
```suggestion
    possible_allergens = set.union(*allergen_map.values())
```

File: 2020/day21/solution.py [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+
+food_re = re.compile(r"(?P<ingredients>.*) \(contains (?P<allergens>.*)\)")
+
+def get_ingredients_and_allergens(line: str) -> tuple[set[str], set[str]]:
+    match = food_re.match(line)
+    ingredients = set(ingredient.strip() for ingredient in match['ingredients'].split(' '))
+    allergens = set(ingredient.strip() for ingredient in match['allergens'].split(','))
+    return (ingredients, allergens)
+
+def produce_allergen_list(allergen_map: dict[str, set[str]]) -> str:
```
Comment (aneeshusa) [inline 2020/day21/solution.py:13]: Couple comments:
- best not to mutate the input dictionary; I did some local edits that a) use a dict comprehension replacing the second loop and parts of the first to do any needed mutations b) replaced `.pop()` with a faster nonmutating alternative.
- return a set of ingredients directly; the caller should be responsible for formatting (e.g. sorting, string combining)
- better names would be helpful in reading (had to rewrite locally with more precise names to understand)

File: 2020/day21/solution.py (hunk 3 above)
Comment (dmmiller) [reply 2020/day21/solution.py]: It does.  Good catch

File: 2020/day21/solution.py (hunk 5 above)
Comment (dmmiller) [reply 2020/day21/solution.py]: That's neat.  I didn't know about that `*` operator, at least as prefix

**Comment by dmmiller** [comment]
@aneeshusa , I updated.  Thanks for the `*` operator, and I made the main loop where I find the unique mapping much more succinct and non mutative using list comprehension.

**Pull request merged**